govcd/catalog.go

```diff
@@ -55,3 +76,352 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
+	}
+
+	var buf bytes.Buffer
+	ovfReader := io.TeeReader(openedFile, &buf)
+
+	request := client.NewRequest(map[string]string{}, "PUT", *ovfUploadUrl, ovfReader)
+	request.Header.Add("Content-Type", "text/xml")
+
+	response, err := client.Http.Do(request)
+	if err != nil {
+		return Envelope{}, err
+	}
+
+	var ovfFileDesc Envelope
+	ovfXml, err := ioutil.ReadAll(&buf)
+	if err != nil {
+		return Envelope{}, err
+	}
+
+	err = xml.Unmarshal(ovfXml, &ovfFileDesc)
+	if err != nil {
+		return Envelope{}, err
+	}
+
+	openedFile.Close()
+	defer response.Body.Close()
+
+	body, err := ioutil.ReadAll(response.Body)
+	log.Printf("[TRACE] Response: %#v\n", response)
+	log.Printf("[TRACE] Response body: %s\n", string(body[:]))
+	log.Printf("[TRACE] Ovf file description file: %#v\n", ovfFileDesc)
+
+	return ovfFileDesc, nil
+}
+
+func findCatalogItemUploadLink(catalog *Catalog) (*url.URL, error) {
+	for _, item := range catalog.Catalog.Link {
+		if item.Type == "application/vnd.vmware.vcloud.uploadVAppTemplateParams+xml" && item.Rel == "add" {
+			log.Printf("[TRACE] Found Catalong link for uplaod: %s\n", item.HREF)
+
+			uploadURL, err := url.ParseRequestURI(item.HREF)
+			if err != nil {
+				return nil, err
+			}
+
+			return uploadURL, nil
+		}
+	}
+	return nil, errors.New("catalog upload url isn't found")
+}
+
+func uploadFile(client *Client, uploadLink, filePath string, offset, fileSizeToUpload int64) (int64, error) {
+	log.Printf("[TRACE] Starting uploading: %s, offset: %v, fileze: %v, toLink: %s \n", filePath, offset, fileSizeToUpload, uploadLink)
+
+	file, err := os.Open(filePath)
+	if err != nil {
+		return 0, err
+	}
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return 0, err
+	}
+
+	defer file.Close()
+
+	request, err := newFileUploadRequest(uploadLink, file, offset, fileInfo.Size(), fileSizeToUpload)
+	if err != nil {
+		return 0, err
+	}
+
+	response, err := client.Http.Do(request)
+	if err != nil {
+		return 0, err
+	}
+	defer response.Body.Close()
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return 0, err
+	}
+	log.Printf("[TRACE] Response: %#v\n", response)
+	log.Printf("[TRACE] Response body: %s\n", string(body[:]))
+
+	if response.StatusCode != http.StatusOK {
+		return 0, errors.New("File " + filePath + " upload failed. Err: " + fmt.Sprintf("%#v", response) + " :: " + string(body[:]) + "\n")
+	}
+	return fileInfo.Size(), nil
+}
+
+func findFilePath(filesAbsPaths []string, fileName string) string {
+	for _, item := range filesAbsPaths {
+		_, file := filepath.Split(item)
+		if file == fileName {
+			return item
+		}
+	}
+	return ""
+}
+
+/** Initiates creation of item and returns ovf upload url for created item. */
+func createItemForUpload(client *Client, createHREF *url.URL, catalogItemName string, itemDescription string) (*url.URL, error) {
+
+	reqBody := bytes.NewBufferString(
+		"<UploadVAppTemplateParams xmlns=\"http://www.vmware.com/vcloud/v1.5\" name=\"" + catalogItemName + "\" >" +
```

Comment:
Can you check the docs to ensure this xmlns tag is still valid? v1.5 may cause problems in Vulcan.  We have dropped support for server API versions below 19.0.

Reply:
According vCloud API 30.0 seems fine: https://code.vmware.com/apis/287/vcloud#/doc/doc/types/UploadVAppTemplateParamsType.html